testbed/testbed/receivers.go

```diff
@@ -0,0 +1,144 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testbed
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/jaegerreceiver"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/opencensusreceiver"
+)
+
+// Receiver allows to receive traces or metrics. This is an interface that must
+// be implemented by all protocols that want to be used in MockBackend.
+// Note the terminology: testbed.Receiver is something that can listen and receive data
+// from Collector and the corresponding entity in the Collector that sends this data is
+// an exporter.
+type Receiver interface {
+	Start(tc *mockTraceConsumer, mc *mockMetricConsumer) error
+	Stop()
+
+	// Generate a config string to place in exporter part of collector config
+	// so that it can send data to this receiver.
+	GenConfigStr() string
+
+	// Return protocol name to use in collector config pipeline.
+	ProtocolName() string
+}
+
+// ReceiverBase implement basic functions needed by all receivers.
+type ReceiverBase struct{}
+
+func (mb *ReceiverBase) Context() context.Context {
+	return context.Background()
+}
+
+func (mb *ReceiverBase) ReportFatalError(err error) {
+	log.Printf("Fatal error reported: %v", err)
+}
+
+// OCReceiver implements OpenCensus format receiver.
+type OCReceiver struct {
+	ReceiverBase
+	receiver *opencensusreceiver.Receiver
+}
+
+func (or *OCReceiver) Start(tc *mockTraceConsumer, mc *mockMetricConsumer) error {
+	addr := "127.0.0.1:56565"
+	var err error
+	or.receiver, err = opencensusreceiver.New(addr, tc, mc)
+	if err != nil {
+		return err
+	}
+
+	// TODO: add metric support.
+	return or.receiver.StartTraceReception(or)
+}
+
+func (or *OCReceiver) Stop() {
+	// TODO: add metric support.
+	or.receiver.StopTraceReception()
+}
+
+func (or *OCReceiver) GenConfigStr() string {
+	// Note that this generates an exporter config for agent.
+	return `
+  opencensus:
+    endpoint: "localhost:56565"`
+}
+
+func (or *OCReceiver) ProtocolName() string {
+	return "opencensus"
+}
+
+// jaegerReceiver implements Jaeger format receiver.
+type jaegerReceiver struct {
+	ReceiverBase
+	receiver      receiver.TraceReceiver
+	jaegerErrChan chan error
+	port          int
+}
+
+const DefaultJaegerPort = 14268
+
+func NewJaegerReceiver(port int) *jaegerReceiver {
+	return &jaegerReceiver{port: port}
+}
+
+func (jr *jaegerReceiver) Start(tc *mockTraceConsumer, mc *mockMetricConsumer) error {
+	jaegerCfg := jaegerreceiver.Configuration{
+		CollectorHTTPPort: jr.port,
+	}
+	var err error
+	jr.receiver, err = jaegerreceiver.New(context.Background(), &jaegerCfg, tc, zap.NewNop())
+	if err != nil {
+		return err
+	}
+
+	jr.jaegerErrChan = make(chan error)
+
+	go func() {
+		for err := range jr.jaegerErrChan {
```

Comment:
I didn't see any code writing to this channel, is it needed?

Reply:
Good catch. This is a remnant from old code. I have no idea what it was supposed to be used for. And I wrote this only a few months ago. Where is my memory? :-)

Reply:
Fixed.